### PR TITLE
Support for strong authentication of time

### PR DESCRIPTION
**Pull request by Bren2010**
- The main change here is to have each log entry contain the time it was created. This provides a stronger authentication of time compared to just signing it into the tree head and prevents certain classes of attacks.
- Introduces the concept of a "maximum lifetime" for log entries. Essentially, if a log entry is too old, users will be able to see that and guide their search away from old portions of the tree, allowing them to be pruned.
- Introduces the concept of a "reasonable monitoring window". This is the opposite of the "maximum lifetime" idea -- it allows users to tell when certain log entries have been added so recently that other users likely haven't had time to vet their contents. When users see that information has been added within the reasonable monitoring window, they maintain extra state about it to prevent/detect potential attacks. Similarly, when information is older than the reasonable monitoring window, it is taken more at face value.

### CODE REVIEW

**Review by thibmeu — COMMENTED** [review]
this is a great addition. I have done a read, not an implementation. only minor comment, no blocker

File: draft-ietf-keytrans-protocol.md [hunk 1]
```diff
@@ -1005,71 +1244,120 @@ that the result equals the root value of the prefix tree.
...
+- For each log entry touched by the algorithm in {{fv-algorithm}}:
+  - The log entry's timestamp.
+  - If the log entry has surpassed its maximum lifetime and is on the frontier,
+    the right child's timestamp.
+  - If it is not the case that the log entry has surpassed its maximum lifetime,
+    is on the frontier, and the log entry's right child has also surpassed its
+    maximum lifetime, then a `PrefixProof` corresponding to a binary ladder in
+    the log entry's prefix tree is provided.
+- If the `PrefixProof` from the first log entry to contain the label-version
+  pair didn't touch the target version, provide a second `PrefixProof` from this
```
Comment (felixlinker) [inline draft-ietf-keytrans-protocol.md:1324]: What does "the label-version pair" refer to? The target? And what does "touch" mean? Doesn't every prefix proof "touch" the target as in it shows inclusion/non-inclusion?

File: draft-ietf-keytrans-protocol.md [hunk 2]
```diff
@@ -1190,23 +1441,54 @@ struct {
...
+  opaque commitment<Hash.Nh>;
+} BinaryLadderStep;
+
 struct {
   FullTreeHead full_tree_head;
 
-  SearchProof search;
-  opaque opening<16>;
+  optional<uint32> version;
+  BinaryLadderStep binary_ladder<0..2^8-1>;
```
Comment (felixlinker) [inline draft-ietf-keytrans-protocol.md:1453]: I don't understand why there is a single binary ladder here. Don't we need multiple (for each node on the path in the implicit binary search tree to the leaf)? And aren't those provided in `search`?

File: draft-ietf-keytrans-protocol.md [hunk 3]
```diff
@@ -1005,71 +1244,120 @@ that the result equals the root value of the prefix tree.
 
 ## Combined Tree {#proof-combined-tree}
```
Comment (felixlinker) [inline draft-ietf-keytrans-protocol.md:1245]: I found this section quite hard to parse, and I don't think I've fully understood it. It started clicking when I arrived at line 1486. You try to define a general data format to accommodate for multiple types of proofs.

Would it maybe help to first list the use cases (i.e., what is included) and then generalize that into one data structure? The use-cases don't explicitly refer to the fields in the `CombinedTreeProof` data structure anyways.

File: draft-ietf-keytrans-protocol.md [hunk 4]
```diff
@@ -1190,23 +1441,54 @@ struct {
...
-   determined that the search was executed incorrectly, abort with an error.
-2. With the candidate root value for the tree, verify the given `FullTreeHead`.
-3. Verify that the commitment in the terminal search step opens to
+1. Compute the VRF output for each version of the label from the proofs in
+   `binary_ladder`.
+2. Verify the proof in `search` as described in {{proof-combined-tree}}.
+3. Compute a candidate root value for the tree from the proof in `inclusion`,
+   the hashes of log entries used in `search`, and any previously-retained full
+   subtrees of the log tree.
+4. With the candidate root value for the tree, verify `FullTreeHead`.
```
Comment (felixlinker) [inline draft-ietf-keytrans-protocol.md:1490]: What does "verify" mean here? I think: check for equality of the hash, no? I think that's short enough s.t. it can be made explicit.

File: draft-ietf-keytrans-protocol.md [hunk 5]
```diff
@@ -1229,15 +1511,18 @@ struct {
 
 If the request passes application-layer policy checks, the Transparency Log adds
 a new label-version pair to the prefix tree, followed by adding a new entry to
-the log tree with the associated value and updated prefix tree root. It returns
+the log tree with an updated timestamp and prefix tree root. It returns
```
Comment (felixlinker) [inline draft-ietf-keytrans-protocol.md:1514]: I think the draft must specify that clients must check that the timestamp the log shows them is sufficiently close to their submission. Otherwise, all that will ever be checked is monotonicity.

File: draft-ietf-keytrans-protocol.md [hunk 6]
```diff
@@ -1257,113 +1542,95 @@ Users initiate a Monitor operation by submitting a MonitorRequest to the
...
+1. Verify that the number of entries in `label_versions` is equal to the number
+   of MonitorLabel structures in `labels` with `rightmost` present. If a
+   MonitorLabel has a `rightmost` field that is not the rightmost distinguished
+   log entry, verify that the corresponding MonitorLabelVersion's `versions`
+   field is not empty.
+2. Verify the proof in `monitor` as described in {{proof-combined-tree}}.
+3. Compute a candidate root value for the tree from the proof in `inclusion`,
+   the hashes of log entries used in `search`, and any previously-retained full
+   subtrees of the log tree.
+4. With the candidate root value for the tree, verify `FullTreeHead`.
```
Comment (felixlinker) [inline draft-ietf-keytrans-protocol.md:1633]: Again, maybe worth pointing out that "verify" means "check for equality."

**Review by felixlinker — APPROVED** [review]
I like the changes and think they're a great addition. I have left some comments, but they are all on details.

File: draft-ietf-keytrans-protocol.md [hunk 7]
```diff
@@ -442,123 +456,102 @@ Index:  0  1  2  3  4  5  6  7  8  9 10 11 12 13
...
+
+## Algorithm {#update-algorithm}
+
+Users retain the following information about the last tree head they've
+observed:
+
+1. The "size" of the log tree, defined as the number of log entries (leaves) it
+   contains.
+2. The root hashes of the "full subtrees" of the log tree. The full subtrees are
+   those which are the largest powers of two possible. This allows the
```
[the remark this answers is not in the document]
Comment (Bren2010) [reply draft-ietf-keytrans-protocol.md:495]: I maybe need a better term. A "full subtree" is any one with 2^n leaves. And then we say we want the smallest set possible of full subtrees that cover the full tree (the largest full subtrees possible).

File: draft-ietf-keytrans-protocol.md [hunk 8]
```diff
@@ -569,96 +562,344 @@ It consists of the following lookups:
...
+aborted. If the user's search would recurse away from the expired portion of the
+tree (to the log entry's right child), the user continues as normal.
+
+When the root and potentially multiple frontier log entries are expired, the
+user skips to the furthest-right expired frontier log entry without receiving
+binary ladders from any of its parents. Similar to the previous case,
+the user is provided with a binary ladder from this log entry. If the user
+determines that its search would recurse to the left (further into the expired
+portion of the tree), it aborts; to the right (into the unexpired portion of the
+tree), it continues.
```
[the remark this answers is not in the document]
Comment (Bren2010) [reply draft-ietf-keytrans-protocol.md:615]: The issue is that, in some cases, we need to provide a proof from the prefix tree in an expired node. Whether or not depends on whether we started at an expired root, or we started at an unexpired node and recursed to an expired one.

File: draft-ietf-keytrans-protocol.md [hunk 9]
```diff
@@ -569,96 +562,344 @@ It consists of the following lookups:
...
+2. If the right timestamp minus the left timestamp is less than the Reasonable
+   Monitoring Window, terminate the algorithm. Otherwise, assert that the given
+   log entry is distinguished.
+3. If the given log entry has a left child in the implicit binary search tree,
+   then recurse to its subtree by executing this algorithm with: the given log
+   entry's left child, the given left timestamp, and the timestamp of the given
+   log entry.
+4. If the given log entry has a right child, then recurse to its right subtree
+   by executing this algorithm with: the given log entry's right child, the
+   timestamp of the given log entry, and the given right timestamp.
```
[the remark this answers is not in the document]
Comment (Bren2010) [reply draft-ietf-keytrans-protocol.md:748]: The implicit tree does not have the property that every non-leaf has both a left and right child.

[the remark this answers is not in the document]
Comment (felixlinker) [reply draft-ietf-keytrans-protocol.md:748]: Ah, makes sense! Still think it'd be beneficial to make it explicit that the algorithm computes a list.

File: draft-ietf-keytrans-protocol.md [hunk 10]
```diff
@@ -569,96 +562,344 @@ It consists of the following lookups:
...
+{{reasonable-monitoring-window}} and {{distinguished-log-entries}} define the
+concept of a distinguished log entry, which is any log entry that label owners
+are required to check for correctness. As a result, users can start their search
+at the rightmost distinguished log entry and only consider new versions which
+have been created since then. The rightmost distinguished log entry will always
+be on the frontier of the log and will never be past its maximum lifetime.
+
+To perform a greatest-version search, the Transparency Log provides the greatest
+version of the label that exists. It then provides a binary ladder, with the
+greatest version as the target, from the rightmost distinguished log entry. If
```
[the remark this answers is not in the document]
Comment (Bren2010) [reply draft-ietf-keytrans-protocol.md:885]: Sorry, what is the contradiction?

File: draft-ietf-keytrans-protocol.md [hunk 11]
```diff
@@ -735,6 +981,16 @@ struct {
 } TreeHeadTBS;
 ~~~
 
+The `max_ahead` and `max_behind` fields contain the maximum amount of time in
+milliseconds that a tree head may be ahead of or behind the user's local clock
+without being rejected. The `reasonable_monitoring_window` contains the
+Reasonable Monitoring Window, defined in {{reasonable-monitoring-window}}, in
+milliseconds. If the Transparency Log has chosen to define a maximum lifetime
```
[the remark this answers is not in the document]
Comment (Bren2010) [reply draft-ietf-keytrans-protocol.md:988]: No, these are durations not timestamps

File: draft-ietf-keytrans-protocol.md [hunk 12]
```diff
@@ -1005,71 +1244,120 @@ that the result equals the root value of the prefix tree.
...
+If a log entry's timestamp is referenced multiple times, it is only added to
+`timestamps` the first time. Additionally, when a user advertises a
+previously-observed tree size in their request, log entry timestamps that the
+user is expected to have retained are excluded from `timestamps`.
+
+The `prefix_proofs` array differs from the `timestamps` array in two important
+ways. First, it is possible for the same log entry to have multiple entries in
+`prefix_proofs` if it is referenced multiple times. Users MUST verify that all
+`PrefixProof` structures corresponding to the same log entry compute the same
+prefix tree root hash. Second, while every log entry with a timestamp in
```
[the remark this answers is not in the document]
Comment (Bren2010) [reply draft-ietf-keytrans-protocol.md:1288]: The timestamps will always be the same, but repeated PrefixProofs may (will) be for different indices

File: draft-ietf-keytrans-protocol.md (hunk 4 above)
Comment (Bren2010) [reply draft-ietf-keytrans-protocol.md:1490]: Will specify this more clearly in a follow-up PR

File: draft-ietf-keytrans-protocol.md (hunk 5 above)
Comment (Bren2010) [reply draft-ietf-keytrans-protocol.md:1514]: Verifying that the rightmost timestamp is sufficiently recent is part of verifying "as if it were a SearchResponse for the most recent version of `label`"
